src/io/input.cpp

```diff
@@ -137,36 +157,55 @@ void gebaar::io::Input::handle_touch_event_down(libinput_event_touch* tev) {
  * @param tev Touch Event
  */
 void gebaar::io::Input::handle_touch_event_up(libinput_event_touch* tev) {
-  touch_swipe_event.up_slots.push_back(std::pair<int, double>(
+  touch_swipe_event.up_slots.push_back(std::pair<size_t, double>(
       libinput_event_touch_get_slot(tev), libinput_event_touch_get_time(tev)));
-  check_multitouch_down_up(touch_swipe_event.up_slots, "up_slots");
-
-  if (touch_swipe_event.up_slots.size() ==
-      touch_swipe_event.down_slots.size()) {
-    std::map<int, std::pair<double, double>>::iterator iter =
-        touch_swipe_event.delta_xy.begin();
-    std::vector<int> swipes;
-    int swipe_type;
-    while (iter != touch_swipe_event.delta_xy.end()) {
-      double x = iter->second.first;
-      double y = iter->second.second;
+  check_multitouch_down_up(touch_swipe_event.up_slots);
+
+  bool a =
+      touch_swipe_event.up_slots.size() == touch_swipe_event.down_slots.size();
+
+  if (a) {
+    std::vector<size_t> swipes;
+    size_t swipe_type;
+    size_t slot;
+    double x;
+    double y;
+    double l;
+    for (auto iter = touch_swipe_event.delta_xy.begin();
+         iter != touch_swipe_event.delta_xy.end(); iter++) {
+      slot = iter->first;
+      x = iter->second.first;
+      y = iter->second.second;
+      l = get_swipe_length(x, y);
+      libinput_device* dev =
+          libinput_event_get_device(libinput_event_touch_get_base_event(tev));
       swipe_type = get_swipe_type(x, y);
 
+      if (touch_swipe_event.fingers == 1) {
+        if (!test_above_threshold(swipe_type, l, dev)) {
+          break;
+        }
+      }
+
+      spdlog::get("main")->debug(
+          "[{}] at {} - {}, slot: {}, swipe-type: {}, length: {}", FN, __LINE__,
+          __func__, slot, config->get_swipe_type_name(swipe_type), l);
+
       if (!swipes.empty() && swipe_type != swipes.back()) {
-        touch_swipe_event.isClean = false;
+        break;
       }
-      spdlog::get("main")->debug("[{}] at {} - {}, swpe-type: {}", FN, __LINE__,
-                                 __func__, swipe_type);
+
       swipes.push_back(swipe_type);
-      iter++;
     }
 
-    if (touch_swipe_event.down_slots.size() !=
-        touch_swipe_event.delta_xy.size()) {
-      touch_swipe_event.isClean = false;
-    }
+    bool b = touch_swipe_event.down_slots.size() ==
+             touch_swipe_event.delta_xy.size();
 
-    if (touch_swipe_event.isClean) {
+    bool c = touch_swipe_event.down_slots.size() == touch_swipe_event.fingers;
+
+    bool d = swipes.size() == touch_swipe_event.fingers;
```

Comment:
Do we need to store these bools `b`, `c`, and `d` ? Can't they be used in the `if` directly ?    
Also can you add a comment of what is this `if` checking (swipe validation) ?  
Also can you please include some debug logging in this area to know if the swipe is being applied.